src/services/bundle.service.js

```diff
@@ -0,0 +1,149 @@
+      response: {
+        status: `${result.status} ${result.statusText}`
+      }
+    };
+    if (result.status === 200 || result.status === 201) {
+      entry.response.location = `/${result.resource.resourceType}/${result.resource.id}`;
+    } else {
+      entry.response.outcome = result.data;
+    }
+    entries.push(entry);
+  });
+
+  bundle.entry = entries;
+  return bundle;
+};
+
+/**
+ * Uploads bundle to server.
+ * @param {Object} request the request object passed in by the user
+ * @param {Object} reply the response object
+ *
+ * @returns {Object} transaction-response/batch-response bundle
+ */
+const uploadTransactionOrBatchBundle = async (request, reply) => {
+  request.log.info('Base >>> Transaction/Batch Bundle Upload');
+  const { resourceType, type, entry: entries } = request.body;
+
+  if (resourceType !== 'Bundle') {
+    reply.code(400).send(new Error(`Expected 'resourceType: Bundle', but received 'resourceType: ${resourceType}'.`));
+  }
+  if (!['transaction', 'batch'].includes(type.toLowerCase())) {
+    reply.code(400).send(new Error(`Expected 'type: transaction' or 'type: batch'. Received 'type: ${type}'.`));
+  }
+
+  if (type.toLowerCase() == 'transaction') {
+    const requestResults = await uploadResourcesFromTxnBundle(entries, reply);
+    const bundle = createResponseBundle(requestResults, reply, 'transaction');
+    request.log.info('Transaction bundle successfully uploaded to server');
+    return bundle;
+  } else {
+    const requestResults = await uploadResourcesFromBatchBundle(entries);
+    const bundle = createResponseBundle(requestResults, reply, 'batch');
+    request.log.info('Batch bundle successfully uploaded to server');
+    return bundle;
+  }
+};
+
+/**
+ * Scrubs transaction bundle entries and uploads each entry to the server.
+ * @param {Array} entries entries from POSTed transaction bundle
+ * @param {Object} reply the response object
+ * @returns array of request results
+ */
+const uploadResourcesFromTxnBundle = async (entries, reply) => {
+  const scrubbedEntries = replaceReferences(entries);
+  const requestsArray = scrubbedEntries.map(async entry => {
+    const { method } = entry.request;
+    return insertBundleResources(entry, method).catch(e => {
+      // send 400 error and stop any further operations from occurring
+      reply.code(400).send(e.message);
+    });
+  });
+  const requestResults = await Promise.all(requestsArray);
+  return requestResults;
+};
+
+/**
+ * Scrubs batch bundle entries and uploads each entry to the server.
+ * @param {Array} entries entries from POSTed batch bundle
+ * @returns array of request results
+ */
+const uploadResourcesFromBatchBundle = async entries => {
+  const scrubbedEntries = replaceReferences(entries);
+  const requestsArray = scrubbedEntries.map(async entry => {
+    const { method } = entry.request;
+    return insertBundleResources(entry, method).catch(() => {
+      // exclude results for resource that produced an error, but
```

Comment:
Out of curiosity, why is an error not thrown for the batch bundles? Is this something that's defined in the spec?

Reply:
From the [spec](https://www.hl7.org/fhir/http.html#brules): "The success or failure of one change SHOULD not alter the success or failure or resulting content of another change...When processing the batch, the HTTP response code is 200 Ok if the batch was processed correctly, regardless of the success of the operations within the Batch. To determine the status of the operations, look inside the returned Bundle. A response code on an entry of other than 2xx (200, 202 etc) indicates that processing the request in the entry failed." 

I actually did not read that second part until now about including the entry in the bundle even if the response code is not 2xx... I'll look into adding that rather than skipping over the failed entries